### PR TITLE
Update exchange resources to support secure application secrets

### PR DESCRIPTION
**Pull request by megha070**
Signed-off-by: megha070 <varshney.megha070@gmail.com>
This PR addresses #491 and aims at updating exchange resources so as to support secure application secrets.

### CODE REVIEW

File: src/main/scala/com/horizon/exchangeapi/ServicesRoutes.scala [hunk 1]
```diff
@@ -259,6 +261,12 @@ trait ServicesRoutes extends JacksonSupport with AuthenticationSupport {
           "defaultValue": "bar"
         }
       ],
+     "secrets": [
+  "secret1": {
+    "type": "string",
+    "description": "string"
+ }
+],
```
Comment (bmpotter) [inline src/main/scala/com/horizon/exchangeapi/ServicesRoutes.scala:269]: This is a minor thing, but please fix the indenting of this section so it is consistent with the others.

File: src/main/scala/com/horizon/exchangeapi/ServicesRoutes.scala [hunk 2]
```diff
@@ -396,6 +406,12 @@ trait ServicesRoutes extends JacksonSupport with AuthenticationSupport {
           "defaultValue": "bar"
         }
       ],
+      "secrets": [
+  "secret1": {
+    "type": "string",
+    "description": "string"
+ }
+],
```
Comment (bmpotter) [inline src/main/scala/com/horizon/exchangeapi/ServicesRoutes.scala:414]: Same comment about indenting

File: src/main/scala/com/horizon/exchangeapi/tables/Services.scala [hunk 3]
```diff
@@ -97,6 +99,7 @@ object ServicesTQ {
   def getMatchHardware(service: String): Query[Rep[String], String, Seq] = rows.filter(_.service === service).map(_.matchHardware)
   def getRequiredServices(service: String): Query[Rep[String], String, Seq] = rows.filter(_.service === service).map(_.requiredServices)
   def getUserInput(service: String): Query[Rep[String], String, Seq] = rows.filter(_.service === service).map(_.userInput)
+  def getUserSecrets(service: String): Query[Rep[String], String, Seq] = rows.filter(_.service === service).map(_.secrets)
```
Comment (bmpotter) [inline src/main/scala/com/horizon/exchangeapi/tables/Services.scala:102]: Another minor thing, but to follow the naming conventions this should be named `getSecrets`

**Review by bmpotter — CHANGES_REQUESTED** [review]

File: src/main/scala/com/horizon/exchangeapi/BusinessRoutes.scala [hunk 4]
```diff
@@ -379,6 +381,17 @@ trait BusinessRoutes extends JacksonSupport with AuthenticationSupport {
...
     }
   ],
+  "secretBinding": [
+     {
+       "serviceOrgid": "string",
+        "serviceUrl": "string",
+        "serviceArch": "amd64",
+        "serviceVersionRange": "x.y.z",
+         "secrets": [
+         { "vaultSecret": "string" }
```
Comment (dabooz) [inline src/main/scala/com/horizon/exchangeapi/BusinessRoutes.scala:391]: is vaultSecret a variable or is it an explicit key/field name? The design says that it is a variable, essentially just a string.

File: src/main/scala/com/horizon/exchangeapi/BusinessRoutes.scala [hunk 5]
```diff
@@ -379,6 +381,17 @@ trait BusinessRoutes extends JacksonSupport with AuthenticationSupport {
       ]
     }
   ],
+  "secretBinding": [
+     {
+       "serviceOrgid": "string",
+        "serviceUrl": "string",
+        "serviceArch": "amd64",
+        "serviceVersionRange": "x.y.z",
```
Comment (dabooz) [inline src/main/scala/com/horizon/exchangeapi/BusinessRoutes.scala:389]: Is a version range acceptable in the value of this field, e.g. "(1.0.0,2.0.0)" ?

**Review by dabooz — CHANGES_REQUESTED** [review]

File: src/main/scala/com/horizon/exchangeapi/BusinessRoutes.scala (hunk 4 above)
Comment (megha070) [reply src/main/scala/com/horizon/exchangeapi/BusinessRoutes.scala:391]: @dabooz It is just a variable.

File: src/test/bash/primedb.sh [hunk 6]
```diff
@@ -8,8 +8,8 @@ if [[ $1 == "-h" || $1 == "--help" ]]; then
 Usage: $0
 
 Required Environment Variables:
-    EXCHANGE_ROOTPW: Exchange root password (Must be similar to what is set in exchange's config.json)
-
+     EXCHANGE_ROOTPW: Exchange root password (Must be similar to what is set in exchange's config.json)
```
Comment (sf2ne) [inline src/test/bash/primedb.sh:11]: I don't think you need to make any changes in `primedb.sh` so you should probably get rid of these.

File: src/main/scala/com/horizon/exchangeapi/tables/Schema.scala [hunk 7]
```diff
@@ -178,13 +178,19 @@ object SchemaTQ {
         sqlu"alter table servicepolicies add column label character varying not null default ''",
         sqlu"alter table servicepolicies add column description character varying not null default ''"
       )
+      case 41 => DBIO.seq(
+        sqlu"alter table patterns add column secretbinding character varying not null default ''"
+      )
+      case 42 => DBIO.seq(
```
Comment (sf2ne) [inline src/main/scala/com/horizon/exchangeapi/tables/Schema.scala:184]: In the future, we try to keep it one DB upgrade step per PR. It's fine for now, but in the future these should be one step.

**Review by sf2ne — CHANGES_REQUESTED** [review]
This PR will be approved once we get the tests passing. I only have one small change I'm requesting in primedb.sh. The rest of my comments are just for future reference.

**Comment by megha070** [comment]
Resolved in another PR!

**Pull request closed without merging**